### ISSUE TITLE
terraform 0.9.7 can't push state to atlas during apply

### ISSUE THREAD

**Issue by clarkperkins** [hashicorp/terraform#15175]
I recently updated to 0.9.7 in atlas, and have not been able to run a successful apply in atlas since then.  The apply always fails with the following error:

```
Md5 The MD5 hash of the state provided does not match what is currently\nknown for the same serial number.
```

### Terraform Version
v0.9.7

### Affected Resource(s)

n/a (everything?)

### Terraform Configuration Files

n/a

### Debug Output
```
Failed to save state: Atlas detected a remote state conflict.

Message: {"errors":["Md5 The MD5 hash of the state provided does not match what is currently\nknown for the same serial number"],"success":false}

Failed to persist state to backend.

The error shown above has prevented Terraform from writing the updated state
to the configured backend. To allow for recovery, the state has been written
to the file "errored.tfstate" in the current working directory.

Running "terraform apply" again at this point will create a forked state,
making it harder to recover.

To retry writing this state, use the following command:
    terraform state push errored.tfstate
```

### Panic Output

n/a

### Expected Behavior

terraform should not fail to push the remote state to atlas

### Steps to Reproduce

1. `terraform apply`

### Important Factoids

The `terraform apply` command is being run on atlas, although I run into the same error locally.  When I do run locally, I can run the `terraform state push errored.tfstate` command and apply again.  That apply fails with the same error, but more infrastructure gets created the second time.  I can keep running those 2 commands over again and eventually get all the infrastructure created, but that is quite frustrating & doesn't allow me to utilize atlas (since I don't have access to the `errored.tfstate` that gets created on the atlas server).

**Comment by clarkperkins** [comment]
Update on this:  I wanted to make sure I hadn't done something stupid, so I downgraded to 0.9.6 and this problem went away.

**Comment by phinze** [comment]
Hi @clarkperkins - thanks for reporting this. I'm seeing it as well in another context. We're actively investigating the problem now!

**Comment by apparentlymart** [comment]
Indeed we're taking a look now, and it looks like we inadvertently introduced a race condition in the code from #14834. I'm trying to figure out now what's the best way to resolve that. In the mean time, remaining on 0.9.6 is the best workaround for now.

**Comment by apparentlymart** [comment]
Hi everyone! Sorry about this regression. (and sorry my PR inadvertently closed this without my noticing.)

We've released a new version 0.9.8 that addresses the problem. We are going to remove 0.9.7 from Terraform Enterprise and will be moving all users of 0.9.7 automatically onto 0.9.8 to get things working. Since Terraform will not accept states from a newer version, downgrading is not easy and so we suggest moving forward onto 0.9.8 to fix this issue.

Sorry again for the regression. We will hold a post-mortem about this tomorrow to see what we can do to avoid issues like this in future.
